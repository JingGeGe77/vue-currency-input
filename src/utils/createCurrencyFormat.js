export default ({ locale, currency }) => {
  const numberFormat = new Intl.NumberFormat(locale, { style: 'currency', currency })
  const str = numberFormat.format(123456)
  const decimalLength = (str.match(/0/g) || []).length
  const decimalSymbol = decimalLength > 0 ? str.substr(str.indexOf('6') + 1, 1) : null
  const prefix = str.substring(0, str.indexOf('1'))
  const negativePrefix = numberFormat.format(-1).substring(0, str.indexOf('1') + 1)
  const suffix = str.substring(str.lastIndexOf(decimalLength > 0 ? '0' : '6') + 1)
  const groupingSymbol = str.substr(str.indexOf('3') + 1, 1)

  return {
    prefix,
    negativePrefix,
    suffix,
    groupingSymbol,
    decimalSymbol,
    decimalLength
  }
}
