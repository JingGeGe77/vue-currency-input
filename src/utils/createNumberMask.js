const nonDigitsRegExp = /\D+/g
const digitRegExp = /\d/
const caretTrap = '[]'

const convertToMask = strNumber => strNumber.split('').map((char) => digitRegExp.test(char) ? digitRegExp : char)

const createNumberMask = (
  {
    prefix = '',
    suffix = '',
    includeThousandsSeparator = true,
    thousandsSeparatorSymbol = ',',
    allowDecimal = true,
    allowNegative = false,
    decimalSymbol = '.',
    decimalLimit = 2,
    integerLimit = null
  } = {}) => {
  const prefixLength = prefix ? prefix.length : 0
  const suffixLength = suffix ? suffix.length : 0
  const thousandsSeparatorSymbolLength = thousandsSeparatorSymbol ? thousandsSeparatorSymbol.length : 0

  return (rawValue = '') => {
    const rawValueLength = rawValue.length
    const indexOfLastDecimal = rawValue.lastIndexOf(decimalSymbol)
    const hasDecimal = indexOfLastDecimal !== -1
    const isNegative = (rawValue[0] === '-') && allowNegative

    let integer, fraction, mask

    if (isNegative) {
      rawValue = rawValue.toString().substr(1)
    }

    if (rawValue.slice(suffixLength * -1) === suffix) {
      rawValue = rawValue.slice(0, suffixLength * -1)
    }

    if (hasDecimal && allowDecimal) {
      integer = rawValue.slice(rawValue.slice(0, prefixLength) === prefix ? prefixLength : 0, indexOfLastDecimal)
      fraction = rawValue.slice(indexOfLastDecimal + 1, rawValueLength)
      fraction = convertToMask(fraction.replace(nonDigitsRegExp, ''))
    } else {
      if (rawValue.slice(0, prefixLength) === prefix) {
        integer = rawValue.slice(prefixLength)
      } else {
        integer = rawValue
      }
    }

    if (integerLimit && typeof integerLimit === 'number') {
      const thousandsSeparatorRegex = thousandsSeparatorSymbol === '.' ? '[.]' : `${thousandsSeparatorSymbol}`
      const numberOfThousandSeparators = (integer.match(new RegExp(thousandsSeparatorRegex, 'g')) || []).length

      integer = integer.slice(0, integerLimit + (numberOfThousandSeparators * thousandsSeparatorSymbolLength))
    }

    integer = integer.replace(nonDigitsRegExp, '')
    integer = integer.replace(/^0+(0$|[^0])/, '$1')

    integer = (includeThousandsSeparator) ? integer.replace(/\B(?=(\d{3})+(?!\d))/g, thousandsSeparatorSymbol) : integer

    mask = convertToMask(integer)

    if (hasDecimal && allowDecimal) {
      if (rawValue[indexOfLastDecimal - 1] !== decimalSymbol) {
        mask.push(caretTrap)
      }

      mask.push(decimalSymbol, caretTrap)

      if (fraction) {
        if (typeof decimalLimit === 'number') {
          fraction = fraction.slice(0, decimalLimit)
        }

        mask = mask.concat(fraction)
      }
    }

    if (prefixLength > 0) {
      mask = prefix.split('').concat(mask)
    }

    if (isNegative) {
      mask = [/-/].concat(mask)
    }

    if (suffix.length > 0) {
      mask = mask.concat(suffix.split(''))
    }

    return mask
  }
}

export default createNumberMask
