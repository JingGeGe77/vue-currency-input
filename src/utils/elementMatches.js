export default (el, selector) => (el.matches || el.msMatchesSelector || el.webkitMatchesSelector).call(el, selector)
