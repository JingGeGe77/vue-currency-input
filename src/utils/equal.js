const equal = (a, b) => {
  if (a === b) {
    return true
  }
  if (a.valueOf !== Object.prototype.valueOf) {
    return a.valueOf() === b.valueOf()
  }
  const keys = Object.keys(a)
  if (keys.length !== Object.keys(b).length) {
    return false
  }
  if (!keys.every(Object.prototype.hasOwnProperty.bind(b))) {
    return false
  }
  return keys.every(key => equal(a[key], b[key]))
}

export default equal
