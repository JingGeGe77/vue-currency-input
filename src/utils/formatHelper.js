export const onlyDigits = (str) => str.replace(/\D+/g, '')

export const removePrefix = (str, prefix) => {
  if (prefix && str.startsWith(prefix)) {
    return str.substr(prefix.length)
  }
  return str
}

export const removeSuffix = (str, suffix) => {
  if (suffix && str.endsWith(suffix)) {
    return str.slice(0, suffix.length * -1)
  }
  return str
}

export const parse = (str, { decimalSymbol, allowNegative = true } = {}) => {
  if (typeof str === 'number') {
    return str
  }
  if (str && typeof str === 'string' && str.trim().length) {
    const negative = str.startsWith('-') && allowNegative
    const numberParts = str.split(decimalSymbol)
    let number = onlyDigits(numberParts[0])
    if (negative) {
      number = '-' + number
    }
    if (numberParts.length === 2) {
      number += '.' + onlyDigits(numberParts[1])
    }
    if (number) {
      number = Number(number)
      return Number.isNaN(number) ? null : number
    }
  }
  return null
}
