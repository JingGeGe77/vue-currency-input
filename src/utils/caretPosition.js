import { count, onlyDigits, removeSuffix } from './formatHelper'

export const setCaretPosition = (el, position) => el.setSelectionRange(position, position)

export const getCaretPositionAfterFormat = (el, inputtedValue, caretPosition) => {
  const { prefix, suffix, decimalSymbol, decimalLength, thousandsSeparatorSymbol } = el.$ci.currencyFormat
  const newValue = el.value
  const decimalSymbolPosition = inputtedValue.indexOf(decimalSymbol) + 1
  let caretPositionFromLeft = inputtedValue.length - caretPosition

  if (newValue === inputtedValue) {
    return caretPosition
  } else if (Math.abs(newValue.length - inputtedValue.length) > 1 && caretPosition <= decimalSymbolPosition) {
    return newValue.indexOf(decimalSymbol) + 1
  } else if (newValue.substr(caretPosition, 1) === thousandsSeparatorSymbol && count(newValue, thousandsSeparatorSymbol) === count(inputtedValue, thousandsSeparatorSymbol) + 1) {
    return newValue.length - caretPositionFromLeft - 1
  } else {
    if (decimalSymbolPosition !== 0 && caretPosition > decimalSymbolPosition) {
      if (onlyDigits(removeSuffix(inputtedValue.substr(decimalSymbolPosition), suffix)).length - 1 === decimalLength) {
        caretPositionFromLeft -= 1
      }
    }
    return el.$ci.options.hideCurrencySymbol
      ? newValue.length - caretPositionFromLeft
      : Math.max(newValue.length - Math.max(caretPositionFromLeft, suffix.length), prefix.length === 0 ? 0 : prefix.length + 1)
  }
}

export const getCaretPositionOnFocus = (el) => {
  let position = el.selectionStart
  const { prefix, thousandsSeparatorSymbol } = el.$ci.currencyFormat
  const { hideCurrencySymbol, hideThousandsSeparatorSymbol } = el.$ci.options
  if (hideCurrencySymbol) {
    position -= prefix.length
  }
  if (hideThousandsSeparatorSymbol) {
    position -= count(el.value.substring(0, el.selectionStart), thousandsSeparatorSymbol)
  }
  return Math.max(0, position)
}
