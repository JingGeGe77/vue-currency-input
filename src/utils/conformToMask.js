import { endsWith, onlyDigits, removeCurrencySymbol, removePrefix, startsWith } from './formatHelper'

const isValidInteger = (integer, { prefix, thousandsSeparatorSymbol }) => integer.replace(prefix, '').match(new RegExp(`^-?(0|[1-9]\\d{0,2}(\\${thousandsSeparatorSymbol}?\\d{3})*)$`))

const isFractionIncomplete = (value, currencyFormat) => {
  const { prefix, decimalSymbol, thousandsSeparatorSymbol } = currencyFormat
  const numberParts = value.split(decimalSymbol)
  return endsWith(value, decimalSymbol) && numberParts.length === 2 && isValidInteger(numberParts[0], {
    prefix,
    thousandsSeparatorSymbol
  })
}

const checkIncompleteValue = (value, previousConformedValue, currencyFormat) => {
  const { prefix, suffix, decimalSymbol, decimalLength } = currencyFormat
  const negative = startsWith(value, '-')
  value = removePrefix(value, '-')
  value = removeCurrencySymbol(value, currencyFormat)
  if (value === '' && negative && previousConformedValue !== `-${prefix}`) {
    return `-${prefix}${suffix}`
  } else if (decimalLength > 0) {
    if (isFractionIncomplete(value, currencyFormat)) {
      return `${negative ? '-' : ''}${prefix}${value}${suffix}`
    } else if (startsWith(value, decimalSymbol)) {
      return `${negative ? '-' : ''}${prefix}0${decimalSymbol}${(onlyDigits(value.substr(1)).substr(0, decimalLength))}${suffix}`
    }
  } else {
    return null
  }
}

const isFractionInvalid = (fraction, numberOfFractionDigits) => fraction.length > 0 && numberOfFractionDigits === 0

export default (value, currencyFormat, previousConformedValue = '') => {
  if (typeof value === 'number') {
    let [integer, fraction] = value.toString().split('.')
    if (fraction) {
      fraction = fraction.substr(0, currencyFormat.decimalLength)
    }
    return {
      conformedValue: Number(`${integer}.${fraction || ''}`),
      fractionDigits: fraction || ''
    }
  } else if (typeof value === 'string') {
    value = value.trim()
    const numberParts = value.split(currencyFormat.decimalSymbol)
    const [integer, ...fraction] = numberParts
    const integerDigits = onlyDigits(integer).replace(/^0+(0$|[^0])/, '$1')
    const fractionDigits = onlyDigits(fraction.join('')).substr(0, currencyFormat.decimalLength)
    const numberOfFractionDigits = fractionDigits.length
    const incompleteValue = checkIncompleteValue(value, previousConformedValue, currencyFormat)
    if (incompleteValue != null) {
      return { conformedValue: incompleteValue }
    }
    if (isFractionInvalid(fraction, numberOfFractionDigits)) {
      return { conformedValue: previousConformedValue }
    }
    let number = integerDigits
    if (startsWith(value, '-')) {
      number = `-${number}`
    }
    if (numberOfFractionDigits > 0) {
      number += `.${fractionDigits}`
    }
    if (number.match(/^-?\d+(\.\d+)?$/)) {
      return {
        conformedValue: Number(number),
        fractionDigits
      }
    } else if (number === '-' && previousConformedValue !== `-${currencyFormat.prefix}`) {
      return { conformedValue: previousConformedValue }
    } else {
      return { conformedValue: '' }
    }
  }
  return { conformedValue: previousConformedValue }
}
