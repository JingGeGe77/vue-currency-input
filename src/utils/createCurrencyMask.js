import { onlyDigits, removePrefix, removeSuffix, startsWith } from './formatHelper'

const digitRegExp = /\d/
const caretTrap = '[]'

const convertToMask = (strNumber) => strNumber.split('').map((char) => digitRegExp.test(char) ? digitRegExp : char)

export default ({ prefix, suffix, thousandsSeparatorSymbol, allowDecimal, allowNegative, decimalSymbol, decimalLimit } = {}) => {
  const prefixLength = prefix ? prefix.length : 0
  const suffixLength = suffix ? suffix.length : 0

  return (rawValue = '') => {
    const indexOfDecimalSymbol = rawValue.indexOf(decimalSymbol)
    const hasDecimal = indexOfDecimalSymbol !== -1
    const negative = startsWith(rawValue, '-') && allowNegative

    rawValue = removePrefix(rawValue, '-')
    rawValue = removePrefix(rawValue, prefix)
    rawValue = removeSuffix(rawValue, suffix)

    let integer, fraction

    if (hasDecimal && allowDecimal) {
      const numberParts = rawValue.split(decimalSymbol)
      integer = numberParts[0]
      fraction = convertToMask(onlyDigits(numberParts[1]))
    } else {
      integer = rawValue
    }

    integer = onlyDigits(integer)
    integer = integer.replace(/^0+(0$|[^0])/, '$1')
    integer = integer.replace(/\B(?=(\d{3})+(?!\d))/g, thousandsSeparatorSymbol)

    let mask = convertToMask(integer)

    if (hasDecimal && allowDecimal) {
      mask.push(caretTrap, decimalSymbol, caretTrap)
      if (fraction) {
        if (typeof decimalLimit === 'number') {
          fraction = fraction.slice(0, decimalLimit)
        }
        mask = mask.concat(fraction)
      }
    }
    if (prefixLength > 0) {
      mask = prefix.split('').concat(mask)
    }
    if (negative) {
      mask = [/-/].concat(mask)
    }
    if (suffixLength > 0) {
      mask = mask.concat(suffix.split(''))
    }
    return mask
  }
}
