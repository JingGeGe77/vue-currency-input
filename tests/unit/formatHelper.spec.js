import { parse, removePrefix, removeSuffix } from '../../src/utils/formatHelper'

describe('formatHelper', () => {
  describe('removePrefix', () => {
    it('removes a prefix', () => {
      expect(removePrefix('abc', 'a')).toBe('bc')
      expect(removePrefix('abc', 'abc')).toBe('')
      expect(removePrefix('abc', '')).toBe('abc')
      expect(removePrefix('abc', 'd')).toBe('abc')
      expect(removePrefix('abc', null)).toBe('abc')
      expect(removePrefix('abc', undefined)).toBe('abc')
    })
  })

  describe('removeSuffix', () => {
    it('removes a suffix', () => {
      expect(removeSuffix('abc', 'c')).toBe('ab')
      expect(removeSuffix('abc', 'abc')).toBe('')
      expect(removeSuffix('abc', '')).toBe('abc')
      expect(removeSuffix('abc', 'd')).toBe('abc')
      expect(removeSuffix('abc', null)).toBe('abc')
      expect(removeSuffix('abc', undefined)).toBe('abc')
    })
  })

  describe('parse', () => {
    it('returns null if the value is empty', () => {
      expect(parse('')).toBeNull()
      expect(parse(' ')).toBeNull()
      expect(parse(null)).toBeNull()
      expect(parse(undefined)).toBeNull()
    })

    it('returns null if the value is invalid', () => {
      expect(parse('-')).toBeNull()
      expect(parse('123e-1')).toBeNull()
      expect(parse('0x11')).toBeNull()
      expect(parse('0b11')).toBeNull()
      expect(parse('0o11')).toBeNull()
      expect(parse('1.2e1', { decimalSymbol: '.' })).toBeNull()
      expect(parse('1.23.4', { decimalSymbol: '.' })).toBeNull()
    })

    it('returns a passed number directly', () => {
      expect(parse(1234)).toBe(1234)
      expect(parse(1.5)).toBe(1.5)
    })

    it('returns the parsed number if the value conforms to the currency format config', () => {
      expect(parse('1234')).toBe(1234)
      expect(parse('1,234,567', { groupingSymbol: ',' })).toBe(1234567)
      expect(parse('$1,234,567', { prefix: '$', groupingSymbol: ',' })).toBe(1234567)
      expect(parse('1234 €', { suffix: ' €' })).toBe(1234)
      expect(parse('-1234')).toBe(-1234)
      expect(parse('-$1234', { prefix: '$' })).toBe(-1234)
      expect(parse('-1234 €', { suffix: ' €' })).toBe(-1234)
      expect(parse('1234.5', { decimalSymbol: null, groupingSymbol: '.' })).toBe(1234.5)
      expect(parse('.5', { decimalSymbol: '.' })).toBe(0.5)
      expect(parse('0.5', { decimalSymbol: '.' })).toBe(0.5)
      expect(parse('1234.50', { decimalSymbol: '.' })).toBe(1234.5)
      expect(parse('$1234.50', { decimalSymbol: '.', prefix: '$' })).toBe(1234.5)
      expect(parse('1234.50 €', { decimalSymbol: '.', suffix: ' €' })).toBe(1234.5)
      expect(parse('1234.5', { decimalSymbol: '.' })).toBe(1234.5)
      expect(parse('1234.00', { decimalSymbol: '.' })).toBe(1234)
      expect(parse('1234.0', { decimalSymbol: '.' })).toBe(1234)
    })

    it('returns null if the value does not conform to the currency format config', () => {
      expect(parse('1234,5', { decimalSymbol: '.' })).toBeNull()
      expect(parse('$1234', { suffix: ' €' })).toBeNull()
      expect(parse('1234 €', { prefix: '$' })).toBeNull()
      expect(parse('1,234,567', { groupingSymbol: '.' })).toBeNull()
    })
  })
})
