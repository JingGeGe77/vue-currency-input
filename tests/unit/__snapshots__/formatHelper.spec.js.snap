// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatHelper getCurrencyFormatConfig returns the correct config 1`] = `
Object {
  "allowDecimal": true,
  "allowNegative": true,
  "decimalLimit": 2,
  "decimalSymbol": ".",
  "prefix": "€ ",
  "suffix": "",
  "thousandsSeparatorSymbol": ",",
}
`;
